unionml/dataset.py

```diff
@@ -34,39 +37,69 @@ def __init__(
         self._random_state = random_state
 
         # default component functions
+        self._loader = self._default_loader
         self._splitter = self._default_splitter
         self._parser = self._default_parser
+        self._parser_feature_key: int = 0  # assume that first element of parser tuple output contains features
         self._feature_loader = self._default_feature_loader
+        self._feature_transformer = self._default_feature_transformer
 
         self._reader = None
         self._reader_input_types: Optional[List[Parameter]] = None
         self._reader_return_type: Optional[Dict[str, Type]] = None
-        self._labeller = None
         self._dataset_task = None
 
     def reader(self, fn=None, **reader_task_kwargs):
+        """Register a reader function for getting data from some external source."""
         if fn is None:
             return partial(self.reader, **reader_task_kwargs)
         self._reader = fn
         self._reader_task_kwargs = reader_task_kwargs
         return fn
 
+    def loader(self, fn):
+        """Register an optional loader function for loading data into memory for model training.
+
+        This function should take the output of the reader function and return the data structure needed
+        for model training.
+        """
+        self._loader = fn
+        return fn
+
     def splitter(self, fn):
-        # TODO: make sure function signature matches expected signature:
-        # - splitter(data, test_size, shuffle, random_state)
+        """Register an optional splitter function that partitions data into training and test sets."""
         self._splitter = fn
         return fn
 
-    def parser(self, fn):
-        # TODO: make sure function signature matches expected signature:
-        # - parser(data, features, targets)
+    def parser(self, fn, feature_key: int = 0):
```

Comment:
Should the type of `feature_key` be more restrictive? I'm concerned about discoverability.

Reply:
what do you suggest?